ships/0039-prune-intermediate-container-images.md

```diff
@@ -0,0 +1,143 @@
+- Mitigate the storage effects of the double push by cleaning up the first container image that was pushed in case post processing results in a second container image.
+
+This enhancement proposal is about the later one.
+
+### Goals
+
+Users will not have two images in their container registry when a double push scenario occurrs.
+
+### Non-Goals
+
+Avoid having two container images in the target registry.
+
+## Proposal
+
+Introduce additional configuration settings in the BuildSpec that enables the possibility to remove the untagged image.
+
+### User Stories
+
+#### Story 1
+
+User can define a setting so that the untagged image from the first push will be pruned.
+
+#### Story 2
+
+User can use Shipwright as-is with the same behavior as before, but end up with an untagged image.
+
+### Implementation Notes
+
+Add `PruneUntaggedImage` to define whether intermediate output images should be pruned. It defaults to `false`, which is the current behavior of doing nothing with the additional image.
+
+```yaml
+apiVersion: shipwright.io/v1beta1
+kind: Build
+metadata:
+  name: buildpack-nodejs-build
+spec:
+  source:
+    type: Git
+    git:
+      url: https://github.com/shipwright-io/sample-nodejs
+    contextDir: source-build
+  strategy:
+    name: buildpacks-v3
+    kind: ClusterBuildStrategy
+  output:
+    image: image-registry.openshift-image-registry.svc:5000/build-examples/taxi-app
+    timestamp: SourceTimestamp
+    pruneUntaggedImage: true
+```
+
+### Test Plan
+
+This feature can only be tested in an E2E test case, since it requires a container registry.
+
+Run a build with default settings and verify that there are two images being created.
+
+Run a build with pruning of untagged images and verify that there is only one image in the registry.
+
+### Release Criteria
+
+n/a
+
+### Risks and Mitigations
+
+## Drawbacks
+
+This is only a mitigtion, there will be still an addtional image in the container registry for a brief moment.
+
+## Alternatives
+
+As touched briefly in the beginning, the alternative would be to rewrite the build in such a way that an additional intermediate container registry can be used. This additional registry doesn't necessarily have to be a temporary registry that is deployed in the build pod or along with the build, but could be a standalone container registry that needs to be configured and acts as a temporary location. However, this approach requires a significant more planning work and additional enhancement proposal to clarify the requirements.
```

Comment:
Conceptually, such an intermediate container registry would also allow to fully use vulnerability scanning for strategy-managed push with failOnFindings set to true. Today, we would not allow such a setting because we cannot prevent pushing to the destination if the strategy does it already.